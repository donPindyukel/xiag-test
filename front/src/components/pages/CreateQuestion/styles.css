.create-question {
  margin: 40px auto;
  max-width: 505px;

  .form-wrap {
    display: flex;
    flex-direction: column;
    justify-content: center;

    input {
      width: 65%;
      @media (width < 768px) {
        font-size: 14px;
      }
    }

    .question-field,
    .answer-wrap {
      display: flex;
      justify-content: space-between;
      padding: 10px 20px;
      border-bottom: 1px var(--borderColor) solid;
      align-items: center;
    }

    .question-field {
      background-color: #eee;

      .question-title {
        text-align: left;
        color: var(--fontColor);
        font-size: 16px;
        font-weight: 700;
      }
    }
  }

  .answers-wrap {
    .answer-wrap {
      font-size: 14px;

      .answer-title {
        color: var(--fontColor);
          font-weight: 700;
      }

      input {
        font-size: 16px;
        @media (width < 768px) {
          font-size: 14px;
        }
      }
    }
  }

  button {
    &:hover {
      box-shadow: none;
    }
  }

  .add-button-wrap {
    border-bottom: 1px var(--borderColor) solid;
    padding: 8px 20px;

    button {
      font-size: 20px;
      padding: 0 10px;
      border: none;
    }
  }

  .start-button-wrap {
    button {
      padding: 0.6rem 2.5rem;
      font-size: 20px;
      margin: 20px 0 0;
      font-family: "Arial", sans-serif;
      &:disabled {
        background: var(--borderColor);
        box-shadow: none;
      }
    }
  }
}
