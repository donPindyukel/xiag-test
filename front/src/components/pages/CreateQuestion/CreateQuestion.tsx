import React from "react";
import { API } from "Services/api";
import { RouteComponentProps } from "react-router";
import "./styles.css";

export interface ICreateQuestionState {
  question: string;
  answers: string[];
  isValidForm: boolean;
  pollId: string;
}

export type ICreateQuestionProps = RouteComponentProps & {};

export class CreateQuestion extends React.Component<
  ICreateQuestionProps,
  ICreateQuestionState
> {
  state: ICreateQuestionState = {
    question: "",
    answers: ["Yes", "No"],
    isValidForm: false,
    pollId: ""
  };

  questionInputHandler = (e: React.ChangeEvent<HTMLInputElement>): void => {
    const question = e.target.value;
    this.setState(
      {
        question
      },
      this.formValidation
    );
  };

  changeAnswerHandler = (index: number) => (
    e: React.ChangeEvent<HTMLInputElement>
  ) => {
    this.state.answers[index] = e.target.value;
    this.setState({
      answers: [...this.state.answers]
    });
    this.formValidation();
  };

  clickAddButtonHandler = () => {
    const answers = [...this.state.answers];
    answers.push("");
    this.setState({
      answers
    });
  };

  clickStartHandler = async () => {
    const { question, answers } = this.state;
    const reqAnswers = answers.map(answer => ({
      title: answer
    }));
    const req = {
      question,
      answers: reqAnswers
    };
    try {
      const data = await API.createQuestion(req);
      this.redirectToPoll(data.data);
    } catch (err) {
      console.log("Error occurred", err);
    }
  };

  redirectToPoll(data: any) {
    const pollId = data._id;
    this.setState(
      {
        pollId
      },
      this.redirectToPollRender
    );
  }

  formValidation = () => {
    const { question, answers } = this.state;
    const isAnswersValid = answers.every(answer => !!answer);
    if (!!question && isAnswersValid) {
      this.setState({
        isValidForm: true
      });
    } else {
      this.setState({
        isValidForm: false
      });
    }
  };

  answerRender = (answer: string, index: number) => {
    return (
      <div className="answer-wrap" key={index}>
        <div className="answer-title">Answer {index + 1}:</div>
        <input
          type="text"
          onChange={this.changeAnswerHandler(index)}
          value={answer}
        />
      </div>
    );
  };

  redirectToPollRender = () => {
    this.props.history.push({
      pathname: `/${this.state.pollId}`
    });
  };

  render() {
    const { question, answers, isValidForm } = this.state;
    return (
      <div className="create-question">
        <div className="form-wrap">
          <div className="question-field">
            <div className="question-title">Question:</div>
            <input
              type="text"
              onChange={this.questionInputHandler}
              value={question}
            />
          </div>
          <div className="answers-wrap">{answers.map(this.answerRender)}</div>
          <div className="add-button-wrap">
            <button onClick={this.clickAddButtonHandler}>+</button>
          </div>
          <div className="start-button-wrap">
            <button disabled={!isValidForm} onClick={this.clickStartHandler}>
              Start
            </button>
          </div>
        </div>
      </div>
    );
  }
}
