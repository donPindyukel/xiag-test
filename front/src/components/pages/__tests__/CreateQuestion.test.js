import React from "react";
import { shallow } from "enzyme";
import toJson from "enzyme-to-json";
import { CreateQuestion } from "../CreateQuestion/CreateQuestion";

describe("<CreateQuestion />", () => {
  let component;

  describe("CreateQuestion", () => {
    beforeAll(() => {
      component = shallow(<CreateQuestion />);
    });

    it("renders the component", () => {
      expect(toJson(component)).toMatchSnapshot();
    });
  });
});