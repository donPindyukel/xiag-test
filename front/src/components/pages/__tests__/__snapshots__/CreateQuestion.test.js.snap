// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<CreateQuestion /> CreateQuestion renders the component 1`] = `
<div
  className="create-question"
>
  <div
    className="form-wrap"
  >
    <div
      className="question-field"
    >
      <div
        className="question-title"
      >
        Question:
      </div>
      <input
        onChange={[Function]}
        type="text"
        value=""
      />
    </div>
    <div
      className="answers-wrap"
    >
      <div
        className="answer-wrap"
        key="0"
      >
        <div
          className="answer-title"
        >
          Answer 
          1
          :
        </div>
        <input
          onChange={[Function]}
          type="text"
          value="Yes"
        />
      </div>
      <div
        className="answer-wrap"
        key="1"
      >
        <div
          className="answer-title"
        >
          Answer 
          2
          :
        </div>
        <input
          onChange={[Function]}
          type="text"
          value="No"
        />
      </div>
    </div>
    <div
      className="add-button-wrap"
    >
      <button
        onClick={[Function]}
      >
        +
      </button>
    </div>
    <div
      className="start-button-wrap"
    >
      <button
        disabled={true}
        onClick={[Function]}
      >
        Start
      </button>
    </div>
  </div>
</div>
`;
